src/lib/sample_sheet.rs

```diff
@@ -0,0 +1,587 @@
+                let key = record[0].to_string();
+                let argument = format!("--{}", key);
+                argv.push(argument);
+
+                // For boolean options, only the presence of a key and the value is ignored.  Thus,
+                // when an empty value is given, assume the argument is a boolean.
+                if !record[1].is_empty() {
+                    argv.push(record[1].to_string());
+                }
+            }
+            line_index += 1;
+        }
+
+        // Build the opts given the arguments.  This is where any unknown command line argument
+        // will raise an error, and illegal value too.
+        match opts.try_update_from(argv.into_iter()) {
+            Ok(()) => Ok(line_index),
+            Err(err) => {
+                let kind = err.kind.as_str().unwrap_or("Unknown error").to_string();
+                let mut args: Vec<String> = vec![];
+                for string in err.info {
+                    args.push(string.split(' ').next().unwrap_or(&string).to_string());
+                }
+                Err(SampleSheetError::DemuxOptionsParsing { kind, args: args.join(", ") })
+            }
+        }
+    }
+
+    /// Reads from a Sample Sheet represented as a vector if `StringRecord`s.  Skips over lines
+    /// optionally until the `[Demux]` section is found, upating the given command line options,
+    /// and then the `[Data]` section, to parse a list of samples.
+    fn from_sample_sheet_string_records(
+        records: &[StringRecord],
+        opts: &Opts,
+    ) -> Result<Self, SampleSheetError> {
+        if records.is_empty() {
+            return Err(SampleSheetError::Empty);
+        }
+
+        let mut line_index = 0;
+
+        // ignore data pre-"[Demux]"
+        while line_index < records.len()
+            && !records[line_index].is_empty()
+            && &records[line_index][0] != "[Demux]"
+        {
+            line_index += 1;
+        }
+
+        // Parse demultiplexing settings
+        let mut opts = opts.clone();
+        let mut line_index = SampleSheet::slurp_demux_opts(records, line_index, &mut opts)?;
+        if records.len() == line_index {
+            // no "[Data]" found
+            return Err(SampleSheetError::NoData);
+        }
+
+        // Parse data
+        line_index += 1; // skip over the "[Data]" line
+        if records.len() == line_index {
+            // no header line in the "[Data]" section
+            return Err(SampleSheetError::NoSamples);
+        }
+        let data_header = &records[line_index];
+        //let data_header: Vec<&str> = records[line_index].into_iter().collect();
+        line_index += 1; // skip over the header
+        let mut ordinal = 1;
+        let mut samples: Vec<SampleMetadata> = vec![];
+        while line_index < records.len() {
+            // allow an empty line
+            // TODO: allow a line with the correct # of columns but all empty lines
```

Comment:
This seems fairly important given what I've seen of sample sheets in the wild.  Perhaps even check to see if all fields are empty/whitespace only.